test/integration/integration_test.cc

```diff
@@ -1796,6 +1796,143 @@ TEST_P(IntegrationTest, Response204WithBody) {
   EXPECT_EQ(0, response->body().size());
 }
 
+// Verifies that a 304 response returns without Transfer-Encoding nor Content-Length header
+TEST_P(IntegrationTest, Response304WithoutTransferEncoding) {
+  initialize();
+  auto tcp_client = makeTcpConnection(lookupPort("http"));
+
+  ASSERT_TRUE(tcp_client->write(
+      "GET /test/long/url HTTP/1.1\r\nHost: host\r\nif-none-match: \"1234567890\"\r\n\r\n"));
+
+  FakeRawConnectionPtr fake_upstream_connection;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  std::string data;
+  ASSERT_TRUE(fake_upstream_connection->waitForData(
+      FakeRawConnection::waitForInexactMatch("\r\n\r\n"), &data));
+  ASSERT_THAT(data, StartsWith("GET"));
+
+  ASSERT_TRUE(fake_upstream_connection->write("HTTP/1.1 304 OK\r\netag: \"1234567890\"\r\n\r\n"));
+
+  tcp_client->waitForData("\r\n\r\n", false);
+  std::string response = tcp_client->data();
+
+  EXPECT_THAT(response, StartsWith("HTTP/1.1 304 Not Modified\r\n"));
+  EXPECT_THAT(response, Not(HasSubstr("transfer-encoding: chunked\r\n")));
+  EXPECT_THAT(response, Not(HasSubstr("content-length")));
+  EXPECT_THAT(response, EndsWith("\r\n\r\n"));
+
+  ASSERT_TRUE(fake_upstream_connection->close());
+  ASSERT_TRUE(fake_upstream_connection->waitForDisconnect());
+  tcp_client->close();
+}
+
+// Verifies same behavior for HEAD request with a 304 response returns without Transfer-Encoding nor
+// Content-Length header
+TEST_P(IntegrationTest, HeadResponse304WithoutTransferEncoding) {
+  initialize();
+  auto tcp_client = makeTcpConnection(lookupPort("http"));
+
+  ASSERT_TRUE(tcp_client->write(
+      "HEAD /test/long/url HTTP/1.1\r\nHost: host\r\nif-none-match: \"1234567890\"\r\n\r\n"));
+
+  FakeRawConnectionPtr fake_upstream_connection;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  std::string data;
+  ASSERT_TRUE(fake_upstream_connection->waitForData(
+      FakeRawConnection::waitForInexactMatch("\r\n\r\n"), &data));
+  ASSERT_THAT(data, StartsWith("HEAD"));
+
+  ASSERT_TRUE(fake_upstream_connection->write("HTTP/1.1 304 OK\r\netag: \"1234567890\"\r\n\r\n"));
+
+  tcp_client->waitForData("\r\n\r\n", false);
+  std::string response = tcp_client->data();
+
+  EXPECT_THAT(response, StartsWith("HTTP/1.1 304 Not Modified\r\n"));
+  EXPECT_THAT(response, Not(HasSubstr("transfer-encoding: chunked\r\n")));
+  EXPECT_THAT(response, Not(HasSubstr("content-length")));
+  EXPECT_THAT(response, EndsWith("\r\n\r\n"));
+
+  ASSERT_TRUE(fake_upstream_connection->close());
+  ASSERT_TRUE(fake_upstream_connection->waitForDisconnect());
+  tcp_client->close();
+}
+
+// Verify content-length is preserved from upstream 304 Response
+TEST_P(IntegrationTest, Response304WithContentLength) {
```

Comment:
I think the two tests below should go into the protocol_integration_test to test the full matrix of upstream/downstream protocols.